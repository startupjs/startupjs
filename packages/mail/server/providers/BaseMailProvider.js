export default class BaseMailProvider {
  constructor (name, send) {
    this.name = name
    this._send = send
  }

  getName () {
    return this.name
  }

  async send () {
    await this._send()
  }
}
