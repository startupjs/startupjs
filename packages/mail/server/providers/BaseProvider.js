// Base provider - interface for all other providers
export default class BaseProvider {
  constructor (params) {
    if (this.instance) return this
    constructor(params)
    this.instance = this._initClient(params)
    return this
  }

  _initClient () {
    throw new Error('[@startupjs/mail] "initClient" not implemented')
  }

  send () {
    throw new Error('[@startupjs/mail] "send" not implemented')
  }
}
