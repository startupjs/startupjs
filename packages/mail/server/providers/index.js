export * from './initProviders'
export { default as BaseMailProvider } from './BaseMailProvider'
