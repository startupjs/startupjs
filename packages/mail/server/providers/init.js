export let providers
export let defaultProvider

/*
config:
{
  Mailgun: {
    instance: require(...),
    options: {
      ...
    }
  }
}
*/

export default function initProviders (options = {}) {
  const {
    defaultProvider: _defaultProvider,
    providers: providersConfig
  } = options

  let _providers = {}
  const providerNames = Object.keys(providersConfig)

  // If no default provider passed then we take
  // first provider from options as default
  defaultProvider = _defaultProvider || providerNames[0]

  for (let providerName of providerNames) {
    _providers[providerName] = new providersConfig[providerName].instance(
      providersConfig[providerName].options
    )
  }

  providers = _providers
}
