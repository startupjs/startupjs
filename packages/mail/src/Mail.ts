import { MailSettings } from './MailSettings';
import { Provider } from './Provider';
import { SendSettings } from './Send';
import { Template } from './Template';

export class Mail {
  private mailSettings: MailSettings;
  constructor(mailSettings: MailSettings) {
    this.mailSettings = mailSettings;

    this.send = this.send.bind(this);
  }

  async send(settings: SendSettings): Promise<any> {
    const availableProviders = this.mailSettings.providers;
    let availableTemplates = this.mailSettings.templates;
    const defaultProviderName = this.mailSettings.default.provider;
    const defaultTemplateName = this.mailSettings.default.template;
    const redefinedTemplates = this.mailSettings.overwrite?.templates || [];

    const targetProviderName =
      settings?.provider || defaultProviderName.getName();
    const targetTemplateName =
      settings?.template || defaultTemplateName.getName();

    let provider: Provider | undefined;
    let template: Template | undefined;

    provider = availableProviders.find(
      availableProvider => availableProvider.getName() === targetProviderName
    );

    if (!provider) {
      throw new Error(
        `[@startupjs/mail] provider: ${targetProviderName} not found.`
      );
    }

    if (Array.isArray(redefinedTemplates) && redefinedTemplates.length > 0) {
      availableTemplates = availableTemplates.map(template => {
        const overwriteTemplate = redefinedTemplates.find(
          item => item.getName() === template.getName()
        );

        if (overwriteTemplate) {
          return overwriteTemplate;
        }

        return template;
      });
    }

    if (availableTemplates.length === 0) {
      throw new Error('[@startupjs/mail] template list is empty.');
    }

    template = availableTemplates.find(
      template => template.getName() === targetTemplateName
    );

    if (!template) {
      throw new Error(
        `[@startupjs/mail] template: ${targetTemplateName} not found.`
      );
    }

    return await provider.send({
      from: settings.from,
      to: settings.to,
      cc: settings.cc,
      bcc: settings.bcc,
      subject: settings.subject,
      inline: settings.inline,
      attachment: settings.attachment,
      tls: settings.tls,
      template: {
        text: await template.getTemplateBody(),
        context: await template.composeData(),
      },
    });
  }
}
