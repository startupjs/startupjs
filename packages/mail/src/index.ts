import express from 'express';
import { MailSettings } from './MailSettings';
import { MailStartupJS } from './MailStartupJS';
import { SendSettings } from './Send';

export { Mail } from './Mail';
export { MailSettings } from './MailSettings';
export { MailStartupJS } from './MailStartupJS';
export { Provider } from './Provider';
export { Send, SendSettings } from './Send';
export { Template } from './Template';

let mail: MailStartupJS | null = null;

export const initMail = (
  settings: MailSettings,
  expressApplication: express.Application,
  apiRootPath: string,
  reInit = false
): void => {
  if (!mail || reInit) {
    mail = new MailStartupJS(settings, expressApplication, apiRootPath);
  }
};

export const sendEmail = async (settings: SendSettings): Promise<void> => {
  if (mail) {
    await mail.send(settings);
  } else {
    throw new Error('[@startupjs/mail] mail service should be init first.');
  }
};
