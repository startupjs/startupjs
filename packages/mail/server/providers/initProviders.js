import _config from '../config'
import BaseMailProvider from './BaseMailProvider'

export default function initProviders (options = {}) {
  const {
    defaultProvider,
    providers
  } = options

  if (!providers) {
    throw new Error(
      '[@startupjs/mail] initProviders: providers is required'
    )
  }

  Object.keys(providers).forEach(provider => {
    if (!(providers[provider] instanceof BaseMailProvider)) {
      throw new Error(`[@startupjs/mail]: Provider ${provider} is not an instance of BaseMailProvider`)
    }
  })

  _config.defaultProvider = defaultProvider || Object.keys(providers)[0]

  _config.providers = providers
}

export function getProvider (name = _config.defaultProvider) {
  if (!_config.providers) {
    throw new Error(
      '[@startupjs/mail] getProvider: initialize ' +
      'library using initMail before getting provider'
    )
  }

  if (!_config.providers[name]) {
    throw new Error(
      `[@startupjs/mail] getProvider: provider: ${name} not found. ` +
      'Initialize it using initMail first.'
    )
  }

  return _config.providers[name || _config.defaultProvider]
}
