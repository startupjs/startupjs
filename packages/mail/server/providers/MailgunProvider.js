import Mailgun from 'mailgun-js'
import BaseProvider from './BaseProvider'

class MailService extends BaseProvider {
  constructor (params) {
    super(params)
    if (this.instance) return this
    const { apiKey, domain } = params
    this.instance = this._initClient({ apiKey, domain })
    return this
  }

  _initClient ({ apiKey, domain }) {
    if (!apiKey) throw new Error('[@startupjs/mail] _initClient: apiKey is required')
    return new Mailgun({ apiKey, domain })
  }

  send (data) {
    try {
      return this.instance.messages().send(data)
    } catch (error) {
      console.log('send email ERR:', error)
      return { error }
    }
  }
}

export default MailService
