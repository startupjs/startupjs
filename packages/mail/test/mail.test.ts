import { Mail } from '../src'
import { Provider1 } from './providers/Provider1'
import { Template1 } from './templates/Template1'

describe('Check mail with', () => {
  it('one provider and one template', async () => {
    const provider1 = new Provider1()
    const template1 = new Template1()

    const mail: Mail = new Mail({
      providers: [provider1],
      templates: [template1],
      default: { provider: provider1, template: template1 }
    })

    expect(
      await mail.send({
        from: 'dmitriy@borodin.site',
        to: 'borodin.d8@gmail.com',
        subject: 'one provider and one template'
      })
    ).toMatchSnapshot()
  })
})
