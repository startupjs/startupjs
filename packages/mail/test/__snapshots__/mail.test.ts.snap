// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check mail with one provider and one template 1`] = `
Object {
  "attachment": undefined,
  "bcc": undefined,
  "cc": undefined,
  "from": "dmitriy@borodin.site",
  "inline": undefined,
  "subject": "one provider and one template",
  "template": Object {
    "context": Object {
      "Template1": Object {
        "user": "Dmitriy",
      },
      "Template2": Object {
        "user": "Oleg",
      },
      "Template3": Object {
        "user": "Lena",
      },
      "Template4": Object {
        "user": "Lada",
      },
    },
    "text": "<!DOCTYPE html>
<html>
  <head>
    <title>%recipient.Template1.title%</title>
  </head>
  <body>
    <h2>Template 1</h2>
    <p>
      Some text... Info: %recipient.Template1.info%
    </p>
    <div>
      <h2>Template 2</h2>
<p>
  Some text... Info: %recipient.Template2.info%
</p>
<p>Template 3 will here</p>
<div>
  <h2>Template 3</h2>
<p>
  Some text... Info: %recipient.Template3.info%
</p>
<p>Template 4 will here</p>
<div>
  <h2>Template 4</h2>
<p>
  Some text... Info: %recipient.Template4.info%
</p>

</div>

</div>

    </div>
    <div>
      <h2>Template 4</h2>
<p>
  Some text... Info: %recipient.Template4.info%
</p>

    </div>
    <div>
      <h2>Template 3</h2>
<p>
  Some text... Info: %recipient.Template3.info%
</p>
<p>Template 4 will here</p>
<div>
  <h2>Template 4</h2>
<p>
  Some text... Info: %recipient.Template4.info%
</p>

</div>

    </div>
  </body>
</html>
",
  },
  "tls": undefined,
  "to": "borodin.d8@gmail.com",
}
`;
